feat:sidecar rewrite

misohu opened the pull request:
This PR covers the rewrite to sidecar. PR covers unit and integration tests.

**Otucomes**

- `unit-kubeflow-dashboard-0` in active state
- `centraldashboard-config` configmap presented in namespace. This cm is used for sidebar population in UI. In case there is cm presented before the charm deployment, the cm will be used instead of creating base one. 
- `profiles crd (admin)`
- `cluterrole` and `clusterrolebinding`

**Quick Dev test (excluding istio)**: run the integration tests and keep the models (check debug logs):
```
tox -e integration -- --model kubeflow --keep-models -vv -s
juju debug-log --replay --include unit-kubeflow-dashboard-0
```

**Full bundle test (including istio)**: 
- run the whole kubeflow localy `juju deploy kubeflow-lite --trust` (dont forgtet [dex configuration steps](https://charmed-kubeflow.io/docs/install))
- replace with current dashboard `juju remove-application kubeflow-dashboard`
- build current charm `charmcraft pack`
- deploy `juju deploy ./*.charm --resource oci-image=$(yq eval '.resources.oci-image.upstream-source' metadata.yaml) --trust`
- dont forget relations `juju relate kubeflow-profiles kubeflow-dashboard; juju relate istio-pilot:ingress kubeflow-dashboard:ingress`
- check the dashboard in browser

**Important**: this charm uses chrome selenium driver for integration tests instead of fiefox

ca-scribner commented on src/charm.py:
```diff
@@ -205,6 +187,74 @@ def _check_kf_profiles(self, interfaces):
+                current_configmap = self.lightkube_client.get(
+                    ConfigMap, name=self._context["configmap_name"]
+                )
+            except Exception as e:
+                self.logger.info(f"ConfigMap not found: {e}. Creating one")
```
As is, we get here for any error that comes up.  Should this catch only the errors raised by lightkube when an object is missing (I think it is something like a 404 error?  You'll need to check)

misohu replied:
Good catch

ca-scribner commented on src/charm.py:
```diff
@@ -205,6 +187,74 @@ def _check_kf_profiles(self, interfaces):
+                self.logger.info(f"ConfigMap not found: {e}. Creating one")
+                self._create_resources()
+            else:
+                self.logger.info(f"ConfigMap found: {current_configmap}. Reusing it")
```
Is there a reason why we do not overwrite any existing configmap?  I'd think having a configmap named the same as what we expect is probably a configuration error (one of: someone made that conflicting cm by mistake, or it was leftover by mistake from a past install).  Was this how the charm handled conflicting cm's before the rewrite?

misohu replied:
Hmm I was thinking it should be naturaly like that :D .... but also the other way makes sense. Previous one was always creating new one. But I fixed it according to your comments

ca-scribner commented on src/charm.py:
```diff
@@ -205,6 +187,74 @@ def _check_kf_profiles(self, interfaces):
 
         return kf_profiles
 
+    def _create_resources(self, resource_type: List[str] = None) -> None:
```
I think a lot of what is done in this overlaps with what we're doing in the [chisme KubernetesResourceHandler](https://github.com/canonical/charmed-kubeflow-chisme/blob/a0a99c9675c0a9948558fb957a18fb601762f67f/src/charmed_kubeflow_chisme/kubernetes/_kubernetes_resource_handler.py#L37).  Have you seen that helper?  If not, take a look.  If it doesn't make sense for the use here, can you explain what doesn't work?

misohu replied:
Firstly I tried ... apply many is not patching resources if failed .... I was thinking to update chisme but it could be breaking change for that function ....

ca-scribner commented on .github/workflows/integrate.yaml:
```diff
@@ -62,6 +62,8 @@ jobs:
             provider: microk8s
             channel: 1.22/stable
             charmcraft-channel: latest/candidate
+      # https://github.com/charmed-kubernetes/actions-operator/issues/27
+      - run: sudo snap refresh charmcraft --channel latest/candidate
```
I think [this is implemented](https://github.com/charmed-kubernetes/actions-operator/pull/28)

misohu replied:
Thank you

ca-scribner commented on .github/workflows/integrate.yaml:
```diff
@@ -70,15 +72,22 @@ jobs:
           # https://github.com/kubeflow/kubeflow/issues/6136
           juju add-model kubeflow
-          sudo apt install -y firefox-geckodriver
+          wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
+          sudo apt install ./google-chrome-stable_current_amd64.deb
+          sudo apt-get install chromium-chromedriver
```
Can we separate this into its own step so it is clear what is environment setup and what is test?

misohu replied:
Thanks

ca-scribner commented on metadata.yaml:
```diff
@@ -20,3 +21,6 @@ requires:
     interface: k8s-service
     schema: https://raw.githubusercontent.com/canonical/operator-schemas/master/k8s-service.yaml
     versions: [v1]
+  sidebar:
+    interface: sidebar
+    optional: true
```
are we implementing the new sidebar as well as rewriting to sidecar in this PR?  What is #38 for?

misohu replied:
Perfect

ca-scribner commented on src/charm.py:
```diff
@@ -176,9 +137,26 @@ def _check_model_name(self):
+                )
+                self._container.restart(self._container_name)
+            except ChangeError as e:
+                self.logger.error(traceback.format_exc())
+                self.unit.status = BlockedStatus("Failed to restart")
+                raise e
```
Have you encountered this case in a full deployment and it done what you expect?  If not, test it see what gets printed to logs and what status is set.  I might be wrong on what will happen, but I think you want either the first two lines or the third, but not both.

What I expect is:
* `self.logger.error()` will print the traceback to juju debug-log
* `self.unit.status = BlockedStatus()` will set the status
* `raise e` (if uncaught by anything above, which I don't see) will cause Juju to print the error message to debug-log (as second time) and set status to `ErrorStatus` (covering over the `BlockedStatus` and requiring the user to intervene with a `juju resolved`)

ca-scribner replied:
If you want to raise `BlockedStatus` for this situation, think about using the same pattern as `_check_container_connection` and others

ca-scribner commented on src/templates/auth_manifests.yaml.j2:
```diff
@@ -0,0 +1,49 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
(this is my sidecar/pebble-ignorance showing) is there no sidecar-native way to grant these permissions?  I'd have expected this to be necessary for odd cases, but not for something standard like this.  @DomFleischmann @DnPlas ?

DnPlas replied:
I'd expect the charm to have permissions granted for most standard k8s resources, but probably not the workload, and also I am not so sure those permissions are extended to custom resources.

misohu replied:
Without the cluster role I am not able to create the profile CRD (e.g. in integration test).

DnPlas replied:
Yeh so it is the latter case: permissions are not extended to the workload to manage custom resources

ca-scribner commented on src/charm.py:
```diff
@@ -28,142 +40,91 @@ def __init__(self, msg: str, status_type=None):
+                        "USERID_PREFIX": "",
+                        "PROFILES_KFAM_SERVICE_HOST": f"{profiles_service}.{self.model.name}",
+                        "REGISTRATION_FLOW": self.model.config["registration-flow"],
+                        "DASHBOARD_LINKS_CONFIGMAP": self.model.config[
+                            "dashboard-configmap"
+                        ],
```
I know this is what was here before in the pod_spec charm, but dashboard-configmap feels like an odd thing to have as a config option.  Especially once you switch to having a dynamically generated sidebar

misohu replied:
I mean for me having the configuration option there makes sense if we dont remove existing configmap .... if we always erase than it make less of a sense .... the configmap change is automatically propagated ... you will see in #38 after I close this PR

ca-scribner commented on src/charm.py:
```diff
@@ -176,9 +137,26 @@ def _check_model_name(self):
+        """Updates the Pebble configuration layer if changed."""
+        current_layer = self._container.get_plan()
+        new_layer = self.get_kubeflow_dashboard_operator_layer(profiles_service)
+        self.logger.info(f"NEW LAYER: {new_layer}")
```
Add a TODO so you don't forget to remove later?

misohu replied:
When what ? sorry I dont get it :D

DnPlas commented on src/charm.py:
```diff
@@ -28,142 +40,91 @@ def __init__(self, msg: str, status_type=None):
-                                },
-                            }
-                        ]
+    def get_kubeflow_dashboard_operator_layer(self, profiles_service: str) -> Layer:
```
In order to do so, and get rid of the `profiles_service` param, you could add an extra property just for that.

misohu replied:
Makes sense thanks :)

DnPlas commented on src/config/sidebar_config.json:
```diff
@@ -0,0 +1 @@
+{"menuLinks": [{"type": "item","link": "/katib/","text": "Experiments (AutoML)","icon": "kubeflow:katib"},{"type": "item","link": "/tensorboards/","text": "Tensorboards","icon": "assessment"},{"type": "item","link": "/jupyter/","text": "Notebooks","icon": "book"},{"type": "item","text": "Experiments (KFP)","link": "/pipeline/#/experiments","icon": "done-all"},{"type": "item","link": "/pipeline/#/ ...
```
Could this be formatted?

misohu replied:
Sadly jinja has some issues here when formatting multi line json .... I was trying it for hour to format correctly :(

misohu commented on src/charm.py:
```diff
@@ -205,6 +187,74 @@ def _check_kf_profiles(self, interfaces):
+        try:
+            self.unit.status = MaintenanceStatus("Creating k8s resources")
+            try:
+                self._create_resources(["auths"])
```
Thanks :)

misohu commented on src/charm.py:
```diff
@@ -205,6 +187,74 @@ def _check_kf_profiles(self, interfaces):
+                self._create_resources()
+            else:
+                self.logger.info(f"ConfigMap found: {current_configmap}. Reusing it")
+                self._create_resources(["profiles"])
```
Good catch

misohu commented on src/charm.py:
```diff
@@ -4,17 +4,29 @@
 import json
 import logging
+import traceback
+from typing import List
```
Thank you :)

misohu commented on src/charm.py:
```diff
@@ -28,142 +40,91 @@ def __init__(self, msg: str, status_type=None):
+        )
+        self.env = Environment(loader=FileSystemLoader("src/templates"))
+        self._name = self.model.app.name
+        self._entrypoint = "npm start"
```
makes sense now :) thanks

misohu commented on src/charm.py:
```diff
@@ -28,142 +40,91 @@ def __init__(self, msg: str, status_type=None):
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "entrypoint of the kubeflow_dashboard_operator image",
+                    "command": f"{self._entrypoint}",
```
Thanks

misohu commented on requirements.txt:
```diff
@@ -1,6 +1,9 @@
 
-ops==1.2.0
+jinja2<3.1
+charmed-kubeflow-chisme==0.0.3
```
Yes I mentioned above ... chisme solves only apply resources not the patch properly :( .... I really tried

misohu replied:
I will remove from req

misohu commented on src/charm.py:
```diff
@@ -28,142 +40,91 @@ def __init__(self, msg: str, status_type=None):
+        self._namespace = self.model.name
+        self.lightkube_field_manager = "lightkube"
+        self._lightkube_client: Optional[Client] = None
+        create_global_resource(
+            group="kubeflow.org", version="v1", kind="Profile", plural="profiles"
```
Perfect. Was searching for it <3

misohu commented on requirements.txt:
```diff
@@ -1,6 +1,9 @@
+jinja2<3.1
+charmed-kubeflow-chisme==0.0.3
+lightkube==0.11.0
+ops==1.3.0
```
removed static version :) thanks

misohu commented on src/charm.py:
```diff
@@ -176,9 +137,26 @@ def _check_model_name(self):
+                self.logger.info(
+                    "Pebble plan updated with new configuration, restarting"
+                )
+                self._container.restart(self._container_name)
```
Good catch forgot to update :)

misohu commented on src/charm.py:
```diff
@@ -176,9 +137,26 @@ def _check_model_name(self):
+        """Updates the Pebble configuration layer if changed."""
+        current_layer = self._container.get_plan()
+        new_layer = self.get_kubeflow_dashboard_operator_layer(profiles_service)
+        self.logger.info(f"NEW LAYER: {new_layer}")
```
seeing the layer in the log was the best part of deployment haha :D  but sure i get it

misohu commented on src/charm.py:
```diff
@@ -205,6 +187,74 @@ def _check_kf_profiles(self, interfaces):
+        """Main entry point for the Charm."""
+        try:
+            self._check_container_connection()
+            self._check_model_name()
```
mhm I mean i would need to create separate handler which will call main handler ... or is there way from event to know if it is install event ? :) Thanks

ca-scribner replied:
I vote check it here just for simplicity.  @dnplas is totally right, but it is cheap and this means we don't need to worry about different logic paths.

misohu commented on src/charm.py:
```diff
@@ -28,142 +40,91 @@ def __init__(self, msg: str, status_type=None):
-            )
+    def _check_container_connection(self):
+        if not self._container.can_connect():
+            raise CheckFailed("Waiting for pod startup to complete", WaitingStatus)
```
Is not there a chcance that after a while this will be ok ? I am not sure here if I get the container states correctly :(

DnPlas replied:
Mmmh, yes this could eventually be okay because you are binding all hooks to the same event handler (`main`), though exceptions you are not handling will be raised if you happen to run into the first line for `_update_layer`, which may cause unexpected behaviours.

misohu replied:
make sense thanks :)

misohu commented on src/charm.py:
```diff
@@ -205,6 +187,74 @@ def _check_kf_profiles(self, interfaces):
+                )
+            except Exception as e:
+                self.logger.info(f"ConfigMap not found: {e}. Creating one")
+                self._create_resources()
```
fixed thanks :)

misohu commented on src/charm.py:
```diff
@@ -205,6 +187,74 @@ def _check_kf_profiles(self, interfaces):
+            return
+        try:
+            self.unit.status = MaintenanceStatus("Creating k8s resources")
+            try:
```
I cant use chisme because it does not implement the patch on apply failure .... sometimes the clusterrolebinding exists before so i need to patch it ... or i was expecting the configmap can be rewritten but now it cant so ..

DomFleischmann replied:
So if I understand this conversation correctly, the main reason we are not using chisme here is due to the `dashboard-configmap` option. We don't want to step on users' toes if they are using that option and pointing to their own config map. I believe this is a bit of an antipattern that we usually discourage in other operators, once we encourage users to "bring their own config" it becomes very hard to maintain the different use cases that users can come up with. Considering also how we are going to overwrite things with the sidebar relation this just becomes a lot more complex than necessary. I believe we should go ahead and remove the configmap option entirely and rely on the operator to be the only one that can create it. This will make our lives a lot easier.

DomFleischmann reviewed (commented):
I really like how these unit tests look. Left my 2 cents on the config map option.

misohu commented:
@DnPlas @DomFleischmann @ca-scribner based on the discussion I rewrote the thing to always rewrite all previous k8s resources (including configmap). Based on this I rewrote the charm to use chisme and I updated the tests :) feel free to add your thoughts <3

DnPlas commented on src/charm.py:
```diff
@@ -28,142 +42,101 @@ def __init__(self, msg: str, status_type=None):
+                context=self._context,
+                logger=self.logger,
+            )
+        load_in_cluster_generic_resources(self._k8s_resource_handler.lightkube_client)
```
Are we using kubeflow-dashboard specific CRs that we need to load? I can't seem to find them, but probably I am missing something.

DnPlas replied:
Nvm, I didn't realise `Profiles` was part of `kubeflow.org`

DnPlas commented on src/charm.py:
```diff
@@ -205,6 +197,44 @@ def _check_kf_profiles(self, interfaces):
+            self._check_container_connection()
+            self._check_model_name()
+            self._check_leader()
+            interfaces = self._get_interfaces()
```
Feels more natural if we add `handle_ingress` here
```suggestion
            interfaces = self._get_interfaces()
            self.handle_ingress(interfaces)
```

misohu replied:
I put it there but I must disagree :) by convention the try execpt block should be arround the code where the exception is expected :) (if possible)

DnPlas commented on src/charm.py:
```diff
@@ -205,6 +197,44 @@ def _check_kf_profiles(self, interfaces):
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            self.logger.error(traceback.format_exc())
+            self.unit.status = BlockedStatus("kubernetes resource creation failed")
```
We might want to `return` here also

misohu replied:
Thanks :)

DnPlas commented on src/charm.py:
```diff
@@ -205,6 +197,44 @@ def _check_kf_profiles(self, interfaces):
+        self.model.unit.status = ActiveStatus()
+
+    def _on_remove(self, event):
+        self.unit.status = MaintenanceStatus("Removing k8s resources")
```
AFAIK, we do not handle status on remove event handlers.

misohu replied:
Pietro suggested to set status in this event for better unittesting.

DnPlas commented on src/charm.py:
```diff
@@ -205,6 +197,44 @@ def _check_kf_profiles(self, interfaces):
+            delete_many(self.k8s_resource_handler.lightkube_client, manifests)
+        except ApiError as e:
+            self.logger.warning(f"Failed to delete resources: {e}")
+        self.model.unit.status = ActiveStatus()
```
Same as above
```suggestion
```

misohu replied:
Pietro suggested to set status in this event for better unittesting.

DnPlas commented on tests/integration/test_charm.py:
```diff
@@ -2,122 +2,140 @@
-    """Workaround for web components breaking querySelector.
+@pytest.mark.asyncio
+async def test_configmap_exist():
+    configmap = Client().get(ConfigMap, CONFIGMAP_NAME, namespace="kubeflow")
```
We could use a pytest fixture for the client
```python
@pytest.fixture
async def lightkube_client():
    lightkube_client = Client(field_manager="test")
    yield lightkube_client
```

misohu replied:
Refactored .. Thanks :)

DnPlas commented on tests/unit/test_operator.py:
```diff
@@ -77,10 +67,152 @@ def test_with_relation(harness):
+class TestCharm:
+    @patch("charm.KubernetesServicePatch", lambda x, y: None)
+    @patch("charm.KubeflowDashboardOperator.k8s_resource_handler", MagicMock())
+    def test_check_leader_failure(self, harness: Harness):
```
This test case's title is a bit misleading, what exactly are we trying to test here? Seems to me we are making sure the expected statuses are set.

DnPlas replied:
After reading the test case below, I guess a good option would be to rename this test case to (or similar):
```suggestion
    def test_check_not_leader(self, harness: Harness):
```

DnPlas commented on tests/unit/test_operator.py:
```diff
@@ -77,10 +67,152 @@ def test_with_relation(harness):
+        harness_with_profiles.charm.on.remove.emit()
+        k8s_resource_handler.assert_has_calls([mock.call.render_manifests()])
+        delete_many.assert_called()
+        assert isinstance(harness_with_profiles.charm.model.unit.status, ActiveStatus)
```
No need (not possible also) to check the status after the charm has been removed

misohu replied:
So after discussion with pietro, he mentioned like the body of that event is some code which may end in failure which should be reflected in the status and that should be the way how to test it ... but not sure now

misohu replied:
The thing is that charm is not removed until the even successfully passes

agathanatasha commented on src/charm.py:
```diff
@@ -28,142 +42,101 @@ def __init__(self, msg: str, status_type=None):
+    @property
+    def _kubeflow_dashboard_operator_layer(self) -> Layer:
+        layer_config = {
+            "summary": "dex-auth-operator layer",
```
need to update summary

misohu replied:
Good catch. Thanks :)

misohu commented on tests/integration/test_charm.py:
```diff
@@ -2,122 +2,140 @@
-    await ops_test.model.add_relation("kubeflow-profiles", charm_name)
+async def test_add_profile_relation(ops_test: OpsTest):
+    await ops_test.model.deploy(PROFILES_CHARM_NAME, channel="latest/edge", trust=True)
+    await ops_test.model.add_relation(PROFILES_CHARM_NAME, CHARM_NAME)
```
Fixed, thanks :)

ca-scribner commented on tests/unit/test_operator.py:
```diff
@@ -77,10 +94,165 @@ def test_with_relation(harness):
+
+
+class TestCharm:
+    @patch("charm.KubernetesServicePatch", lambda x, y: None)
```
An optional suggestion - there's no right or wrong answer...

This patch is applied to every test function, and you never need to use the `Mocker` it creates.  You can move this `@patch("...")` up to decorate `TestCharm`, and remove it from all the test functions.  
You could almost do the same thing to the `k8s_resource_handler` patch, too, except you occasionally need the Mocker.  Maybe there's a clever way to address that, but I don't know it.

misohu replied:
Having it on TestCharm may lead into some unexpected behavior when sb will be rewriting it (if she forgets about the class scope of the patch). I was thinking a lot about it. This seemed to be most readable thing. But I am open for discussion.

ca-scribner replied:
Yep fair point.  Thanks!

ca-scribner commented on src/charm.py:
```diff
@@ -28,142 +42,101 @@ def __init__(self, msg: str, status_type=None):
+        self._name = self.model.app.name
+        self._service = "npm start"
+        self._container_name = "kubeflow-dashboard"
+        self._container = self.unit.get_container(self._name)
```
I am, I think unjustifiably, a little little worried that `self.unit.get_container(self._name)` might return a copy of this object rather than a pointer to the object.  If that is the case, we'd effectively be making a copy here and then whenever we use it, it could be stale.  I don't see any obvious signs that this will be a copy in the code, but I didn't see any contract saying it would be a live reference either.  

Since this is used only in the `self.container` property, what about refactoring to make that property:
```
@property
def container(self):
  return self.unit.get_container(self._name)
```

Yes, the charm will have to invoke the `get_container` func a few extra times, but this way you don't depend on how they're passing the data back.

Also, if you know the return either can never change and is always a live reference, then ignore this.  Ignore this too if this is a common pattern I just don't know

misohu replied:
We use this patter in training/katib/dex (all sidecar charms). But it does not mean its good. But I see similar patter in Jons kubernets dashboard charm (although he calls it multiple times in the code instead of assigning it to class param). I vote for tech topic :). But its a valid worry I like the thought process.

ca-scribner replied:
Good plan.  I'll make a tech topic, can you add links to any charms I miss?  thx!

ca-scribner commented on src/charm.py:
```diff
@@ -205,6 +197,46 @@ def _check_kf_profiles(self, interfaces):
+            self.unit.status = BlockedStatus("kubernetes resource creation failed")
+            return
+        self.handle_ingress(interfaces)
+        kf_profiles = list(kf_profiles.get_data().values())[0]
```
Could this be extracted into a helper?  I know it is a one liner, but I don't think this line immediately says "get data from the kf_profiles interface" to someone who is unfamiliar with the relation.

misohu replied:
welll I dont know so the name will be `get_data_from_profile_interface` ? I think thats what the line literally says in Python except the interface :) or do I get it wrong ?

ca-scribner replied:
Yeah, something named like that is fine with me.

The part that I'd like to hide is the stuff around the get data, eg:
```
list(__.values())[0]
```
To me, that isn't obviously just a "get our data" line but instead could be something complicated.  To really know what we're doing here without entering a debugger, a reader needs to know in detail how SerializedDataInterface works (specifically, what an interface will return).  I think it would be helpful to hide the actual unpacking

We could also just have a comment above this line saying it, but imo a self-commenting function name is simpler.  Then there's a natural place (the docstring of the helper) to explain that we're unpacking data from that relation

ca-scribner reviewed (commented):
Overall this is in a great state.  I really like the thorough and easy to read tests

ca-scribner commented on src/charm.py:
```diff
@@ -176,9 +149,24 @@ def _check_model_name(self):
-            # We can't do anything useful when not the leader, so do nothing.
             raise CheckFailed("Waiting for leadership", WaitingStatus)
 
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer if changed."""
+        current_layer = self.container.get_plan()
+        new_layer = self._kubeflow_dashboard_operator_layer
+        self.logger.debug(f"NEW LAYER: {new_layer}")
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self.container.add_layer(self._container_name, new_layer, combine=True)
+            try:
+                self.logger.info(
+                    "Pebble plan updated with new configuration, replaning"
+                )
+                self.container.replan()
+            except ChangeError:
+                raise CheckFailed("Failed to replan", BlockedStatus)
```
afaict, this helper is reusable boilerplate.  I've just reviewed the same code in the [kserve sidecar charm](https://github.com/canonical/kserve-operators/pull/2/files#diff-ad55a839047964c277b40ec9dea61ff441e4a7bb59b06de9d159673f6d84ef75R78).  Let's refactor this to be an encapsulated, reusable helper and put it in chisme

(note: I left the same comment in the other charm, too)

misohu replied:
Can we open a task for this (this one is already way too broad IMO) :)

DomFleischmann replied:
Fully agree, opened a jira ticket for it next sprint, also mentioned in the ticket that we should update kserve and dashboard once it is done. Lets not block this PR on it though.

misohu commented on tests/integration/test_charm.py:
```diff
@@ -2,122 +2,146 @@
-
-        driver.get_screenshot_as_file(f"/tmp/selenium-{request.node.name}.png")
+@pytest.mark.asyncio
+async def test_configmap_exist(lightkube_client: Client):
```
Good catch ... fixed :)

misohu commented on tests/unit/test_operator.py:
```diff
@@ -77,10 +94,165 @@ def test_with_relation(harness):
+        assert harness.charm.model.unit.status == WaitingStatus(
+            "Waiting for leadership"
+        )
+        assert (
+            "status_set",
+            "waiting",
+            "Waiting for leadership",
+            {"is_app": False},
+        ) in harness._get_backend_calls()
```
Removing ... It was also strange to me but saw it some charms so I wanted to be cool

misohu commented on tests/unit/test_operator.py:
```diff
@@ -77,10 +94,165 @@ def test_with_relation(harness):
+        assert harness.charm.model.unit.status != WaitingStatus(
+            "Waiting for leadership"
+        )
+        assert (
+            "status_set",
+            "waiting",
+            "Waiting for leadership",
+            {"is_app": False},
+        ) not in harness._get_backend_calls()
```
same as above

ca-scribner reviewed (approved):
Looks great!  Gj
